Implement SushiTee's scoreboard

Dune-jr opened the pull request:
So here is the current state of the @SushiTee's scoreboard:

![image](https://user-images.githubusercontent.com/355114/48676243-50d20c00-eb64-11e8-938e-0116a39b1083.png)

I moved the scoreboard a little down compared to @SushiTee's original proposal because of overlaps with the HUD. 

There's a bunch of text in hud.cpp that should probably be moved up (at the same place as "Red team wins!") or disabled. I'm making the pull request already because I'd like people to test/discuss/suggest changes already.

Flags and non-team gametypes seem to work fine.

Dune-jr committed:
Merged scoreboard from #1171

Dune-jr committed:
Scoreboard fixes & improvements, moved the scoreboard up, just below the HUD. Improve display of spectators

Fisico commented:
thanks for you work dune... looks very good

Dune-jr commented:
You should rather thank @Zatline and @SushiTee :)

oy commented:
Country flags for the players have to be added. Besides that it looks great 👍

Dune-jr commented:
Whoops, I knew I forgot something.

Zatline commented:
disable the HUD all togheter if scorebord is tabed, many games do this it's the most cleanest approuch imo it simple and saves a lot of trouble, u not tabing to look at the HUD don't u? get rid of it ;)
we had already aligned the scorebord perfectly to fit the menu pls don't realign!

Dune-jr commented:
@Zatline Really, I don't see what alignment could be broken, vertically. There is enough room in the bottom afaik.

Zatline commented:
@Dune-jr a Scoreboard is a menu element in it's own a quite substantial one in fact, 
for consistency it has to follow the dimensional rules of menus, the alignment has to be the same.

![ezgif-2-015ffb06197d](https://user-images.githubusercontent.com/344608/48679933-4cbee200-eb96-11e8-9888-4e7b13ef6b37.gif)

Dune-jr commented:
@Zatline I see your point.

However, two things:

- the menus cannot be open at the same time as the scoreboard, right now
- the layout ingame menus is bound to be very likely changed

Removing the HUD would be fine but I would rather not do it until it is necessary. I don't think it's done this way in other games very often?

Zatline commented:
@Dune-jr a lot of games are disabling the HUD when tabing scorebord and yes the scorebord should be tabable when in the default menu where nothing currently sits, also it is very commen that scorbords have a locked size, both hud disable and fixed size are industry standards.

Sonix- commented:
Id still want to see my health while checking scoreboard

Zatline commented:
that doesn't make much sense hmm

oy commented:
@Dune-jr  Don't move it down too much. When there're 48 spectators it shouldn't overlap with the chat typing area ;)

Dune-jr commented:
Moved it up again. Original was 73, I had moved it to 103, reverted to 83, right below the HUD.

Menus cannot be opened simultaneously so there is no alignment problem with that. I think there will be room for 48 (64, rather?) spectators (couldn't test though).

oy commented:
Yeah 64 might be better ;) For spectators it's enough to just show the name+clan.

fokkonaut commented:
> For spectators it's enough to just show the name+clan.

we dont need the clan for spectators, the name alone is enough. the clan can be looked for in the players tab in the esc menu

oy commented:
Looks better with clan and it's part of the identity. But it the 2 should be tied together.

fokkonaut commented:
> Looks better with clan and it's part of the identity. But it the 2 should be tied together.

for me it doesnt look better and imagine 63 people on spec, chaos like on 0.6?

oy commented:
Pretty much the same if the clan is written or not. But with clantag it's easier to identify someone within the list. And people might circumvent it by setting their name to name+tag in case it's not shown.

nheir commented:
I suggest not to align player name to `Spectators (%d)`. The space below may be useful later ;)

![screenshot_2018-11-19_22-15-58](https://user-images.githubusercontent.com/2434994/48735520-ea72e980-ec48-11e8-957b-83ec521f79e1.png)

Dune-jr commented on src/game/client/components/scoreboard.h:
```diff
@@ -7,24 +7,38 @@
+ 		CPlayerStats();
+		void Reset();
+	};
+	CPlayerStats m_aPlayerStats[16];
```
Thanks for spotting that one out!

Dune-jr commented:
I agree, @nheir.

Not sure how clans should be displayed. Those empty parenthesis don't look too nice. Maybe remove them altogether when the player has no clan?

oy commented:
@Dune-jr yeah.
Maybe just write the clan behind the name and use a lower alpha on them?

Dune-jr commented:
> @Dune-jr yeah.
> Maybe just write the clan behind the name and use a lower alpha on them?

Great idea, I'll try to do just that.

oy commented:
Yeah better.

quadCoreZ commented:
isn’t it better to block the warmup messages while scoreboard is opened?

Dune-jr commented:
So the scoreboard handles 16 players just fine, even in non-team mode and with the chat area:

![screenshot_2018-11-20_16-44-57](https://user-images.githubusercontent.com/355114/48784816-b6003b80-ece3-11e8-9f6c-fdea3d5e5e98.png)

However, there is no way to make it display 48 spectators. I propose to just cap the spectators display to three lines, and then just add a `...` for now (or a scrollbar).

Dune-jr commented:
@quadCoreZ I think there could be a lot of improvements on those client-side displays, yeah, especially on which context they appear. I would like to propose to unify them a bit too, and that would help with #1694. But that is another issue.

Dune-jr commented:
@nheir That would surely do it, yes.

Zatline commented:
Score should actually be placed first always! As this is the most relevant information.

- Score / Tee / Name / Clan / (Kill / Death) / Country Flag / Ping

![2b12fef788c3b188cf5a85833892a3fb](https://user-images.githubusercontent.com/344608/48788373-fb743700-ecea-11e8-9632-cd77cf6e211e.png)

Dune-jr commented:
The score position was changed by @SushiTee, probably on purpose to put it next to K/D :/

Dune-jr commented:
Here's what it gives:
![screenshot_2018-11-20_19-43-28](https://user-images.githubusercontent.com/355114/48795311-beb13b80-ecfc-11e8-9222-4323921d0565.png)

oy commented:
Looks odd - the gameplay information should stick together.

Zatline commented:
> The score position was changed by @SushiTee, probably on purpose to put it next to K/D :/
> 
> Might want to revert it, I don't know?

no it was put there becouse my [first concept](https://gyazo.com/f35fd0771a8b18e6e41e40bd105c87d0) had it there, I myself kinda did that and it was a mistake and I wanted it to change later, which why I made this other concept but that got never changed the project died right there.

Zatline commented:
why K/D needs to get rid of

- prevents from playing the objectiv a lot of people care more about their ratio than playing the objective.  In fact peopel may stop playing the objectiv all entierly and only focus on killing, a phenomenon that affects all games that activly or inactivly embrace K/D.
- gameplay could become more campy and try hard becouse of that, now I know that tw is not very noob frendly and that skilled player won't get noobtubed so easy but it's still a concern.
- make space for adding back flags

how K/D could be handeled differently
- K/D could still be implemented later in a much more detailed Statsbord showing everything, 
easy accassable by going in the ingame menus clicking a button.
- make it optional, keep the importance low, like seen in the concept by having 50% trancparancy and purposely put it away from the score, make it a feature.

1stDecadeTW commented:
> why K/D needs to get rid of
> 
> * prevents from playing the objectiv a lot of people care more about their ratio than playing the objective.  In fact peopel may stop playing the objectiv all entierly and only focus on killing, a phenomenon that affects all games that activly or inactivly embrace K/D.
> * gameplay could become more campy and try hard becouse of that, now I know that tw is not very noob frendly and that skilled player won't get noobtubed so easy but it's still a concern.
> * make space for adding back flags
> 
> how K/D could be handeled differently
> 
> * K/D could still be implemented later in a much more detailed Statsbord showing everything,
>   easy accassable by going in the ingame menus clicking a button.
> * make it optional, keep the importance low, like seen in the concept by having 50% trancparancy and purposely put it away from the score, "It's not a bug, it's a feature".

Pause menu (ESC) could display "STATS" of the actual game in the game tab or maybe in the player tab?

Dune-jr commented:
> Pause menu (ESC) could display "STATS" of the actual game in the game tab or maybe in the player tab?

The additional statistics scoreboard is already planned: #158 (and #1556 kinda)

For now, we need a functional scoreboard that can be displayed at any time. I think K/D is very fit for such a scoreboard, and it's such a basic thing I believe its benefits massively outweight the inconvenients @Zatline rightfully raised (stats are implemented in a lot of client mods anyway).

Now if the general consensus goes in that direction, I will remove them.

1stDecadeTW commented:
A dashboard with game and individual player information shown in tables and graphics would be awesome :D 

For example, something like this: 
![screenshot_1](https://user-images.githubusercontent.com/44511917/48810158-645cae80-eced-11e8-919b-ffea7ed33fa4.jpg)

It could include not only flag returns, but dominion of the flag, team attack, defenses, max velocity of the player, etc. It needs to be well designed

1stDecadeTW commented:
A dashboard with game and players information displayed in tables and graphics would be awesome :D

It would not only include returned flags, K/D, etc, but also killing spree, domain of the flag, team attack and defense, ratings at the end of each game (best attacker, defender?), etc... it needs to be well designed.

For example, something like this: 
![screenshot_1](https://user-images.githubusercontent.com/44511917/48810746-5d836b00-ecf0-11e8-9fbb-b289f8f76c4a.jpg)

oy commented on src/game/client/components/scoreboard.cpp:
```diff
@@ -103,50 +142,117 @@ void CScoreboard::RenderSpectators(float x, float y, float w)
+
+	bool NoTitle = pTitle? false : true;
+
+	// count players
```
No need for calculation: 'm_pClient->m_GameInfo.m_aTeamSize[Team]' has it stored, but would add a check on top to leave the function if the team isn't red or blue.

Dune-jr replied:
I added an assert instead. There really should be no other team.

Dune-jr replied:
I would add an assert instead, there really shouldn't be any other team.

oy commented on src/game/client/components/scoreboard.cpp:
```diff
@@ -103,50 +142,117 @@ void CScoreboard::RenderSpectators(float x, float y, float w)
 	}
 	else
 	{
-		if(m_pClient->m_Snap.m_SpecInfo.m_Active && m_pClient->m_Snap.m_SpecInfo.m_SpectatorID >= 0 && m_pClient->m_Snap.m_SpecInfo.m_SpectatorID < MAX_CLIENTS &&
```
This line was correct. 'SPEC_FREEVIEW' is 0 and used for SpecModes.

Dune-jr replied:
Changed. I keep the `m_pClient->m_Snap.m_paPlayerInfos[m_pClient->m_Snap.m_SpecInfo.m_SpectatorID]` check right?

oy commented on src/game/client/components/scoreboard.cpp:
```diff
@@ -342,80 +469,131 @@ void CScoreboard::OnRender()
+		}
 		else if(m_pClient->m_Snap.m_pGameDataTeam)
 		{
 			const char *pRedClanName = GetClanName(TEAM_RED);
```
These lines are there twice, move it outside the if clauses. And check if the clan name is empty there and assign "Red/Blue team".

Dune-jr replied:
> And check if the clan name is empty there and assign "Red/Blue team".

I don't think that's a good idea because we need to keep the `str_format` for later, for the `Localize`:

```
				if(pRedClanName)
					str_format(aText, sizeof(aText), Localize("%s wins!"), pRedClanName);
```

oy replied:
There currently is:
-Localize("%s wins!")
-Localize("Red team") <- pRedTeamName
-Localize("Red team wins!") <- this feels redundant
  = str_format(aBuf, sizeof(aBuf), Localize("%s wins!"), pRedTeamName);

Dune-jr replied:
You're right.

oy commented on src/game/client/components/scoreboard.cpp:
```diff
@@ -342,80 +469,131 @@ void CScoreboard::OnRender()
+	Graphics()->MapScreen(0, 0, Width, 400*3.0f);
+	if(m_pClient->m_Snap.m_pGameData && (m_pClient->m_GameInfo.m_GameFlags&GAMEFLAG_TEAMS) && m_pClient->m_Snap.m_pGameDataTeam)
+	{
+		const char *pRedClanName = GetClanName(TEAM_RED);
```
redundant, see previous comment

oy commented on src/game/client/components/scoreboard.cpp:
```diff
@@ -342,80 +469,131 @@ void CScoreboard::OnRender()
+
+			if(m_pClient->m_Snap.m_pGameDataTeam->m_TeamscoreRed > m_pClient->m_Snap.m_pGameDataTeam->m_TeamscoreBlue)
+			{
+				if(pRedClanName)
```
redundant, see previous comment

oy commented on src/game/client/components/scoreboard.cpp:
```diff
@@ -342,80 +469,131 @@ void CScoreboard::OnRender()
-
-				float w = TextRender()->TextWidth(0, 86.0f, aText, -1);
-				TextRender()->Text(0, Width/2-w/2, 39, 86.0f, aText, -1);
+				if(pBlueClanName)
```
redundant, see previous comment

oy commented on src/game/client/components/scoreboard.cpp:
```diff
@@ -342,80 +469,131 @@ void CScoreboard::OnRender()
+		if(pMsg->m_Victim != pMsg->m_Killer)
+			m_aPlayerStats[pMsg->m_Killer].m_Kills++;
+	}
+	else if(MsgType == NETMSGTYPE_SV_CLIENTDROP && Client()->State() != IClient::STATE_DEMOPLAYBACK)
```
should be:
`else if(MsgId == NETMSGTYPE_SV_CLIENTINFO && Client()->State() != IClient::STATE_DEMOPLAYBACK)
	{
		CNetMsg_Sv_ClientInfo *pMsg = (CNetMsg_Sv_ClientInfo *)pRawMsg;

		if(!m_pClient->m_aClients[pMsg->m_ClientID].m_Active)
		{
				return;
		}
		ResetPlayerStats(pMsg->m_ClientID);
	}`

Dune-jr replied:
done

oy reviewed (commented):
Maybe we should remove the Kill Death columns and put the country flags there instead. Then it should be good.

Dune-jr committed:
Some code fixes

Dune-jr committed:
Fix NumPlayers not counting the right team, add an assert

Sonix- commented:
Dont forget the ready_state indicator for each player

Dune-jr committed:
Red team wins and Blue team wins now using %s team wins for localization

LordSk committed:
Country flags

Dune-jr commented:
Props to @LordSk to help finish the PR while I'm busy.

> Dont forget the ready_state indicator for each player

Is this feature fully implemented, @Sonix-?

I feel like scoreboard integration should be done when everything is properly done.

Sonix- commented:
> Props to @LordSk to help finish the PR while I'm busy.
> 
> > Dont forget the ready_state indicator for each player
> 
> Is this feature fully implemented, @Sonix-?
> 
> I feel like scoreboard integration should be done when everything is properly done.

I dont know, there's a setting in controls and warmup + game starts depend on it in a way; Imo.

Dune-jr commented:
I didn't manage to find how to pause the server. Either way, there is already something to indicate ready state from @SushiTee's code, even though it should be improved:

https://github.com/teeworlds/teeworlds/blob/6b5f2d8a31ade430f04f3ba77cc98dc255e473b1/src/game/client/components/scoreboard.cpp#L281-L283

LordSk committed:
ready state green color, increase chat width

oy committed:
blocked non-country flags. closes #1721

oy committed:
changed default background map to heavens. closes #1566
